src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -2921,6 +2923,185 @@ public void visitTypeCast(JCTypeCast tree) {
         result = tree;
     }
 
+    /**
+     * All the exactness checks between primitive types that require a run-time check
+     * are in `java.lang.runtime.ExactConversionsSupport`. Those methods are in the
+     * form `ExactConversionsSupport.<S>_<T>` where both `S` and `T` are primitive
+     * types and correspond to the runtime check that will be executed when the program
+     * needs to check whether a certain value (that is passed as a parameter) can be
+     * converted to T without loss of information.
+     * Rewrite instanceof if expr : Object and Type is primitive type
+     *
+     * According to the Table 5.5-A. Casting conversions to primitive types in the JLS this
+     * conversion is permitted under a casting context, and employs a narrowing reference conversion
+     * followed by an unboxing conversion. As a result the only check that needs to be done is an
+     * `instanceof` to `Float`. The unboxing conversion from `Float` to `float` is always exact
+     * so if `v` is `Float` then `v` will be `instanceof float` too (without further checks).
+     *
+     *   Object v = ...
+     *   if (v instanceof float)
+     *   =>
+     *   if (v instanceof Float)
+     *
+     * Rewrite instanceof if expr : wrapper reference type
+     *
+     *   Integer v = ...
+     *   if (v instanceof float)
+     *   =>
+     *   if (let tmp$123 = v; tmp$123 != null && ExactConversionsSupport.intToFloatExact(tmp$123.intValue()))
+     *
+     * Rewrite instanceof if expr : primitive
+     *
+     *   int v = ...
+     *   if (v instanceof float)
+     *   =>
+     *   if (let tmp$123 = v; ExactConversionsSupport.intToFloatExact(tmp$123))
+     *
+     * More rewritings:
+     *
+     * - If the `instanceof` check is unconditionally exact rewrite to true.
```

Comment:
Note: the same effect can be obtained, w/o rewriting (and perhaps more simply) by detecting the case of an unconditionally exact "instanceof" test, and giving the instanceof expression a constant boolean type (with value "true"). This can be done in Attr. Then, `Gen` will shortcircuit the expression, as required.